lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,1441 @@
+# *-* coding: utf-8 *-*
+"""
+This modules provides methods for generating and collecting ngrams.
+
+The methods allow to collect different kind of subsequences, such as standard
+ngrams (preceding context), skip ngrams with both single or multiple gap
+openings (both preceding and following context), and positional ngrams (both
+preceding and following context).
+"""
+
+from __future__ import division
+import math
+import operator
+import pickle
+import random
+from collections import defaultdict, Counter
+from itertools import chain, combinations, product
+from functools import partial
+from lingpy.sequence.smoothing import smooth_dist
+
+# Global padding symbol, shared across all functions/class-methods.
+_PAD_SYMBOL = '$$$'
+_ELM_SYMBOL = '###'
+
+def _seq_as_tuple(sequence):
+    """
+    Internal function for automatically converting a string sequence to a
+    tuple, if needed.
+
+    Parameters
+    ----------
+    sequence: list or str
+        The sequence that shall be converted into an iterable.
+
+    Returns
+    -------
+    out: tuple
+        A tuple of the sequence.
+    """
+
+    # We first check for datatype and then for a space, as the first test is
+    # faster (and evaluation is lazy).
+    if isinstance(sequence, str) and ' ' in sequence:
+        return tuple(sequence.split(' '))
+
+    return tuple(sequence)
+
+
+def _accumulate(iterable, func=operator.add):
+    """
+    Return running totals.
+    
+    This implementation replaces itertools.accumulate for compatibility
+    with Python 2.7.
+    """
+    # accumulate([1,2,3,4,5]) --> 1 3 6 10 15
+    # accumulate([1,2,3,4,5], operator.mul) --> 1 2 6 24 120
+    it = iter(iterable)
+    try:
+        total = next(it)
+    except StopIteration:
+        return
+    yield total
+    for element in it:
+        total = func(total, element)
+        yield total
+
+def _random_choices(population, weights=None, cum_weights=None, k=1):
+    """
+    Return a `k` sized list of elements chosen from `population` with
```

Comment:
If this function is in standard from version 3.6, but not in 2.7, we should still keep it, but couldn't we add a 3.6 test that would return the standard function?

One more point: if we put this in a utility module, like suggested by @xrotwang, we'd also document it with sphinx, so again, the first sentence should not fill two lines, only one.

Reply:
Agree about keeping: remember that the issue is not only py2 vs. py3, but even old versions of py3 (the Python3 version in the stable long-support Ubuntu, for example, is 3.5 and they rejected proposals for moving forward).

As for using the standard function, I'd prefer not to: code that checks for versions is always ugly in my opinion, and we don't run into any performance issues here (I tested it with some hundreds of thousands random points). This also guarantees that the same seed will produce the same output in the same machine when either py2 or py3 are used.

I've already moved it to utils, I'll fix the docstring.